### PR TITLE
Jerry/test metrics

### PR DESCRIPTION
**Pull request by jkgenser**
Add tests for metrics based on simulated data and refactor the sequence count function to support exact and superset.

### CODE REVIEW

**Comment by madisonmay** [comment]
LGTM!

**Review by madisonmay — APPROVED** [review]

**Comment by jkgenser** [comment]
Going to add macro- and support-weighted f-score as well as class-level f-score (which macro f-score is a function of)

**Comment by jkgenser** [comment]
I took @madisonmay's advice and also refactored the API for these functions a bit to take `span_type` rather than passing in functions. The `model_f1` function also has an API similar to sklearn as it allows passing in `average=option`. Left existing more verbose functions for backwards compatibility.

File: finetune/util/metrics.py [hunk 1]
```diff
@@ -120,61 +121,122 @@ def sequence_labeling_token_counts(true, predicted):
...
+        FN = len(counts["false_negatives"])
+        TP = len(counts["correct"])
+        recall = calc_recall(TP, FN)
+        precision = calc_precision(TP, FP)
+        results[cls_]["support"] = FN + TP
+        results[cls_]["f1-score"] = calc_f1(recall, precision)
+    return results
+
+
+def model_f1(true, predicted, span_type="token", average=None):
```
Comment (madisonmay) [inline finetune/util/metrics.py]: `sequence_f1`?

File: finetune/util/metrics.py [hunk 2]
```diff
@@ -120,61 +121,122 @@ def sequence_labeling_token_counts(true, predicted):
...
+    """
+    F1-scores per class
+    """
+    count_fn = get_seq_count_fn(span_type)
+    class_counts = count_fn(true, predicted)
+    results = {}
+    for cls_, counts in class_counts.items():
+        results[cls_] = {}
+        FP = len(counts["false_positives"])
+        FN = len(counts["false_negatives"])
```
Comment (madisonmay) [inline finetune/util/metrics.py:191]: One final request -- could we call this 'true_positives'?  I think "correct" was a terrible naming convention I wrought upon metrics.py a long ago

File: finetune/util/metrics.py [hunk 3]
```diff
@@ -120,61 +121,122 @@ def sequence_labeling_token_counts(true, predicted):
...
+        results[cls_]["support"] = FN + TP
+        results[cls_]["f1-score"] = calc_f1(recall, precision)
+    return results
+
+
+def model_f1(true, predicted, span_type="token", average=None):
+    """
+    If average = None, return per-class F1 scores
+    """
+    if average == "micro":
```
Comment (madisonmay) [inline finetune/util/metrics.py:205]: The flow of this function is kind of confusing:

I think you could do something like:

```python3
if average == "micro":
    return micro_f1(true, predicted, span_type)

f1_by_class = per_class_f1(true, predicted, span_type)
f1s = [f1s[cls_]['f1-score'] for cls_ in f1_by_class]
supports = [f1s[cls_]['support'] for cls_ in f1_by_class]

if average == 'weighted':
    return np.average(np.average(f1s, weights=np.array(supports))
if average == "macro":
    return np.average(f1s)
else:
    return f1s
```

File: finetune/util/metrics.py [hunk 4]
```diff
@@ -185,7 +247,27 @@ def sequences_overlap(true_seq, pred_seq):
...
+            pred_seq["start"] == true_seq["start"]
+            and pred_seq["end"] == true_seq["end"]
+        )
+
+
+def sequence_superset(true_seq, pred_seq):
+    """
+    Boolean return value indicates whether or predicted seq is a superset of target
+    """
+    return (
```
Comment (madisonmay) [inline finetune/util/metrics.py]: Slightly funky formatting for some reason

Comment (jkgenser) [reply finetune/util/metrics.py]: fixed.

File: finetune/util/metrics.py (hunk 3 above)
Comment (jkgenser) [reply finetune/util/metrics.py:205]: cleaned up.

File: finetune/util/metrics.py (hunk 2 above)
Comment (jkgenser) [reply finetune/util/metrics.py:191]: updated

File: finetune/util/metrics.py (hunk 1 above)
Comment (jkgenser) [reply finetune/util/metrics.py]: updated

**Review by madisonmay — APPROVED** [review]

**Pull request merged**